### PR TITLE
Ep33/35 table plans

### PR DESCRIPTION
**Pull request by dwyerse**
View any event as an admin, and click the yellow button labeled 'Table Plan'. Shows a list of tables with the ability to add, delete and edit tables.

### CODE REVIEW

File: views/table.ejs [hunk 1]
```diff
@@ -0,0 +1,151 @@
+<head>
+	<title>Events</title>
```
Comment (SeanDurban) [inline views/table.ejs]: Should be title of page

File: views/table.ejs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	<!-- Bootstrap CSS -->
+	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
+</head>
+<br>
+<a href="/event/view/<%=event_id%>"><button class="btn btn-primary">Return to event</button> </a>
+<body class="container">		
+	<br>
+	<div style="text-align: center;font-size:2em;"><%=eventName%> - Table Plan</div>
+	<br>
+	<% if (succ && succ.length > 0) { %>
```
Comment (SeanDurban) [inline views/table.ejs]: For consistency would be best to have flash messages at the top as they show below page header

File: views/table.ejs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+				<%}%>
+			</div>			
+			
+		
+		</div>
+			<%}else{%>
+			<div style="text-align:center">-No tables allocated-</div>
+			<%}%>
+			<div class="row">
+				<button type="button" style="margin-bottom:10px;"class="btn btn-warning" data-toggle="modal" data-target="#tableModal">Create Table</button>
```
Comment (SeanDurban) [inline views/table.ejs]: This is my personal preference so ignore if theres a reason for it: There's likely to be multiple tables and they could be quite large because of this the create table button ends up a long way down the page, so might be better suited above the tables

File: routes/table.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	var data = req.body;
+	for(var item in data){
+		if(item.split('-')[0] == 'label'){
+			Labels.push(req.body[item]);
+		}
+		else{
+			Seats.push(req.body[item]);
+		}
+	}
+	var tableObj = new Table({tableNumber:req.params.tableNumber,tableLabels:Labels,size:req.params.labelAmount,seatLabels:Seats,eventId:req.params.event_id});
```
Comment (SeanDurban) [inline routes/table.js]: Your labelAmount param is always 0 for me, this means the size gets reset in DB to 0 everytime you edit the table in any way

File: views/table.ejs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+											<% for(var j=0; j < table[i].tableLabels.length; j++) { %>
+												
+													<input type="name" name="label-<%=j%>" class="labels<%=j%> form-control" value="<%= table[i].tableLabels[j]%>" name="label" rows="5" required><br>												
+												
+											<%}%> 
+											</div>	
+											<% if(table[i].tableLabels.length ==0){%>
+												<div id="noLabels<%=i%>" style="margin:10px; text-align: center;">-No labels-</div>
+											<%}%>
+											<div data-table="<%=i%>" class="addLabel btn btn-secondary btn-sm">Add label</div>
```
Comment (SeanDurban) [inline views/table.ejs]: As mentioned in message: add label functionality is not as I would expect and cannot be recovered from

File: routes/table.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+		tableMapper.findTablesByEventId(req.params.event_id,function(err,table){
+			res.render('table', {table:table,event_id:req.params.event_id,eventName:eventName, succ: req.flash('succ')});
+		});	
+	});	
+});
+
+router.post('/edit/:event_id/:tableNumber/:labelAmount',isLoggedIn, isAdminUser, function(req, res) {
+	
+	var Labels = [];
```
Comment (SeanDurban) [inline routes/table.js]: no camelCase... Ken would be appalled :OO

File: routes/table.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+var isLoggedIn = require('../config/utils').isLoggedIn;
+var isAdminUser = require('../config/utils').isAdminUser;
+
+router.get('/:event_id',isLoggedIn, isAdminUser, function(req, res) {
+	eventMapper.findEventBy_event_id(req.params.event_id,function(err,result){
+
+		var eventName = result.title;
+
+		tableMapper.findTablesByEventId(req.params.event_id,function(err,table){
+			res.render('table', {table:table,event_id:req.params.event_id,eventName:eventName, succ: req.flash('succ')});
```
Comment (SeanDurban) [inline routes/table.js]: flash err messages never shown

File: mappers/tableMapper.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+			return callback(err);
+		}
+		else{
+			return callback(err,res);
+		}
+	});	
+}
+
+function editTable(eventId,tableNumber,tableObj,callback){
+	Table.findOne({eventId:eventId,tableNumber:tableNumber}, function(err,res){
```
Comment (SeanDurban) [inline mappers/tableMapper.js]: This works fine but just for the future there's a mongoose function to exactly this: findOneAndUpdate() - avoids boilerplate of resetting the fields to save

File: routes/table.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+var Table = require('../models/table');
+var tableMapper = require('../mappers/tableMapper');
+var eventMapper = require('../mappers/eventMapper');
+var isLoggedIn = require('../config/utils').isLoggedIn;
+var isAdminUser = require('../config/utils').isAdminUser;
+
+router.get('/:event_id',isLoggedIn, isAdminUser, function(req, res) {
+	eventMapper.findEventBy_event_id(req.params.event_id,function(err,result){
+
+		var eventName = result.title;
```
Comment (SeanDurban) [inline routes/table.js]: if they attempt to lookup illegal/invalid tables such as table/999 and 999 doesn't exists then the server crashes. Since result is null we're attempting to get result.title. Simple condition here to redirect if result is null would avoid the crash

File: test/test_tableMapper.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
+var assert = require('assert');
+var mapper = require('../mappers/tableMapper');
+var mongoose = require('mongoose');
+const APP_DB = 'mongodb://127.0.0.1/eventplanner_db';
```
Comment (SeanDurban) [inline test/test_tableMapper.js]: Overall this test case is quite hard to follow since all the values are hardcoded, would be more readable if a test table object was defined like other unit tests.

File: test/test_tableMapper.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+describe('tableMapper testing suite', function() {
+	//After all tests reconnect to APP_DB
+	before(function(){
+		mongoose.connect(APP_DB);
+	});
+
+	it('should create a new table', function(done) {
+		mapper.createTable(12,[],5,['a','b','c','d','e'],1,function(err,res){
```
Comment (SeanDurban) [inline test/test_tableMapper.js]: We really shouldn't be creating testing data with production data.  In production its highly likely we'll have an event with event_id, if they have 12 tables then the 12th table is corrupted and deleted

File: test/test_tableMapper.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+		var l;
+
+		mapper.findTablesByEventId(1, function(err2,result) {
+			l = result.length;
+			
+			mapper.deleteTable(1,11,function(err){
+				assert.equal(err,null);
+				mapper.findTablesByEventId(1, function(err2,res) {
+					assert.equal(err2,null);
+					assert.equal(res.length,l-1);
```
Comment (SeanDurban) [inline test/test_tableMapper.js]: This assertion fails if there's already >= 12 tables in event 1: 

1) tableMapper testing suite
       should delete table:

      Uncaught AssertionError [ERR_ASSERTION]: 11 == 12
      + expected - actual

      -11
      +12

File: test/test_tableMapper.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+		mapper.createTable(12,[],5,['a','b','c','d','e'],1,function(err,res){
+			assert.equal(err,null);
+			assert.equal(res.tableNumber,12);
+			done();
+		});
+	});
+
+	it('should find tables by eventId', function(done) {
+		mapper.findTablesByEventId(1,function(err,result){
+			assert.notEqual(result[0].tableNumber,0);
```
Comment (SeanDurban) [inline test/test_tableMapper.js]: What exactly is this unit test testing? That at least one table has been created?

Comment (SeanDurban) [reply test/test_tableMapper.js]: Bit confused! Since result[0] will always be table number =1 if it exists. But not necessarily finding all tables

Comment (dwyerse) [reply test/test_tableMapper.js]: Should have been result.length != 0
